### ISSUE TITLE
CR - max-len: {ignoreRegExpLiterals: true} is too permissive

### ISSUE THREAD

**Issue by refack** [eslint/eslint#9059]
**Tell us about your environment**

* **ESLint Version:** `4.3.0`
* **Node Version:** `master`
* **npm Version:** `5.3.0`

**What parser (default, Babel-ESLint, etc.) are you using?**
N/A

**Please show your full configuration:**
https://github.com/nodejs/node/blob/master/.eslintrc.yaml
https://github.com/nodejs/node/blob/master/lib/.eslintrc.yaml

```yaml
max-len: [error, {code: 80,
                  ignoreRegExpLiterals: true,
                  ignoreUrls: true,
                  tabWidth: 2}]
```
**What did you do? Please include the actual source code causing the issue.**

```js
01: function fatalError(e) {
02:   if (typeof e.stack === 'string') {
03:     process._rawDebug(e.stack);
04:   } else {
05:     const o = { message: e };
06:     Error.captureStackTrace(o, fatalError);
07:     process._rawDebug(o.stack);
08:   }
09:  if (process.execArgv.some((e) => /^--abort[_-]on[_-]uncaught[_-]exception$/.test(e))) {
10:    process.abort();
11:   }
12:   process.exit(1);
13: }
```

**What did you expect to happen?**
L09 should error.
Even though we have `ignoreRegExpLiterals: true` such cases should not be allowed. It's not the RegExp that is super long, it's the multiple expressions in that single line that caused it to be too long.
I understand it's not a straightforward problem to solve, but IMHO it could be minimized with some heuristics.

**What actually happened? Please include the actual, raw output from ESLint.**
Nothing (i.e. lint passed)

**Comment by not-an-aardvark** [comment]
Thanks for creating an issue. In https://github.com/eslint/eslint/issues/8189, there was some discussion about changing the behavior of `ignoreRegExpLiterals` to subtract the length of regex literals from the total line length, rather than ignoring lines with those literals entirely. Would that satisfy your use case?

**Comment by refack** [comment]
I'll think about it (too early now 😴)
PTAL at https://github.com/nodejs/node/pull/14607 IMHO it turned out nice.

**Comment by DonkeyAfterDark** [comment]
Subtracting the string from the character count makes more sense.

**Comment by refack** [comment]
I think replacing `len(/RegExp/)` with `min(n, len(/RegExp/))` where `n` is a configurable number defaulting to `max-len.code * 3 / 4` would be best.
That is the RegExp literal should be at least 75% of the line for it to be ignored. If there are more than 25% other chars most probably it could be refactored to a different line.

**Comment by ljharb** [comment]
That's a really arbitrary number, and I strongly disagree with it.

The only thing that makes sense to me is if "ignore x" is enabled, then the length of x shouldn't count towards the length of the line.

**Comment by refack** [comment]
> That's a really arbitrary number, and I strongly disagree with it.

If we remove the arbitrary default, we could have something like:
```yaml
max-len: [error, {code: 80,
                  ignoreRegExpLiterals: 60,
                  ignoreUrls: true,
                  tabWidth: 2}]
```
* `true` - no change in semantics
* n - Ignore `RegExp`s longer than 60 in lines counting them as 60

---

My negative case is:
https://github.com/nodejs/node/blob/d731369b1d0822d5a766612a8a6ef3416aa9d580/lib/async_hooks.js#L74
```js
if (process.execArgv.some((e) => /^--abort[_-]on[_-]uncaught[_-]exception$/.test(e))) {
```
There is no reason to allow this, when
```js
const evenWithAVeryExplicitName =  /^--abort[_-]on[_-]uncaught[_-]exception$/;
if (process.execArgv.some((e) => evenWithAVeryExplicitName.test(e))) {
```
is a perfectly readable mitigation

**Comment by refack** [comment]
Or use some other parameter of line complexity, like "only ignore if there are 5 or less token in the line"

**Comment by refack** [comment]
Or "ignore only if literal is more than N% of line characters"
(sorry for the brainstorm drip)

In hindsight, this actually could simply be a separate rule - `no-long-literals-in-complex-lines`

**Comment by ljharb** [comment]
I'm suggesting that an ignore of `true` means "ignore it" - meaning, disregard its length - I'm not suggesting any non-boolean value be allowed there.

Measuring complexity is imo way out of scope of max-len (which is itself often used as a poor proxy for complexity).

**Comment by refack** [comment]
@ljharb agreed.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
